pkgs/applications/audio/open-stage-control/default.nix

```
@@ -1,29 +1,26 @@
-{ pkgs, stdenv, lib, fetchFromGitHub, makeBinaryWrapper, makeDesktopItem, copyDesktopItems, nodejs, electron, python3, ... }:
+{ lib, buildNpmPackage, fetchFromGitHub, makeBinaryWrapper, makeDesktopItem, copyDesktopItems, electron, python3, ... }:
 
-let
-  nodeComposition = import ./node-composition.nix {
-    inherit pkgs nodejs;
-    inherit (stdenv.hostPlatform) system;
-  };
-in
-
-nodeComposition.package.override rec {
+buildNpmPackage rec {
   pname = "open-stage-control";
-  version = "1.18.3";
+  version = "1.20.0";
 
   src = fetchFromGitHub {
     owner = "jean-emmanuel";
     repo = "open-stage-control";
     rev = "v${version}";
-    hash = "sha256-AXdPxTauy2rMRMdfUjkfTjbNDgOKmoiGUeeLak0wu84=";
+    hash = "sha256-XgwlRdwUSl4gIRKqk6BnMAKarVvp291zk8vmNkuRWKo=";
   };
 
-  strictDeps = true;
+  patches = [
+    # Use generated package-lock.json since upstream does not provide one in releases
+    ./package-lock.json.patch
+  ];
```

```suggestion
  postPathc = ''
    # Use generated package-lock.json since upstream does not provide one in releases
    ln -s ${./package-lock.json} package.json
  '';
```
we don't need to wrap the file in a patch

---

Unfortunately we do because otherwise the npm dep fetcher won't have the lock file. It's in the buildNpmPackage but I'm on mobile and can't link it at the moment

---

Well it was in the code, not the documentation in a diagnostic: https://github.com/NixOS/nixpkgs/blob/67f352e497f55a26404f334ead855f3b29ea76cb/pkgs/build-support/node/fetch-npm-deps/default.nix#L122

But basically only `patches` will make it into the `fetchNpmDeps` call, not any phase hooks: https://github.com/NixOS/nixpkgs/blob/67f352e497f55a26404f334ead855f3b29ea76cb/pkgs/build-support/node/build-npm-package/default.nix#L32-L36

I had it written originally like you have, and I can for sure confirm that it does not work without wrapping it in a patch (unfortunately -- it feels a bit hacky but so is a project that doesn't distribute a package-lock.json in releases I guess)

---

patches should call postPatch unless overwriten. I think it is failing because https://github.com/NixOS/nixpkgs/blob/3ca855084a6a016c5d5e361a544025a1d860afdd/pkgs/build-support/node/fetch-npm-deps/default.nix#L115 should be -e .

---

That is probably a good idea to change!

It also didn't work in this case, though, because `buildNpmPackage` does not pass `postPatch` to the `fetchNpmDeps` call (only `patches`) and I can't override the deps fetcher manually with the way it is structured (not without forgoing `buildNpmPackage` and manually inlining all of the relevant code and adding hooks and such in this package)

Also the diagnostic message suggested using a patch file so I figured that was the most supported path for the issue where upstream does not provide a package-lock.json

---

We should change it to also accept postPatch because wrapping a new file in a patch is a bit ugly and diffs are horrible displayed on github.